File: 1002-find-common-characters/1002-find-common-characters.ts
function findFreq(word: string): Map<string, number> {
  const freq = new Map<string, number>();
  for (const char of word) {
    freq.set(char, (freq.get(char) ?? 0) + 1);
  }
  return freq;
}

function commonChars(words: string[]): string[] {
  const result: string[] = [];
  const freqs: Map<string, number>[] = [];
  words.forEach((word) => freqs.push(findFreq(word)));
  const base = freqs[0];

  base.forEach((val, key) => {
    let isCommon = true;
    let min = val;

    for (const freq of freqs) {
      if (!freq.has(key)) {
        isCommon = false;
        break;
      }
      min = Math.min(min, freq.get(key)!);
    }

    if (isCommon) {
      for (let i = 0; i < min; i++) {
        result.push(key);
      }
    }
  });

  return result;
}

File: 1002-find-common-characters/1002-find-common-characters.rs
impl Solution {
    pub fn common_chars(words: Vec<String>) -> Vec<String> {
        if words.is_empty() {
            return vec![];
        }

        let mut words = words.clone();
        let mut result = Vec::new();
        let base = words.get(0).unwrap().clone();

        for c in base.chars().into_iter() {
            let mut is_all_contains = true;

            let mut i = 0;
            while i < words.len() {
                let word = words.get_mut(i).unwrap();

                if !word.contains(c) {
                    is_all_contains = false;
                    break;
                }

                *word = word.replacen(c, "", 1);

                i += 1;
            }

            if is_all_contains {
                result.push(String::from(c));
            }
        }

        return result;
    }
}
